src/compiler/types.ts

```diff
@@ -3761,15 +3777,37 @@ namespace ts {
 
     // Indexed access types (TypeFlags.IndexedAccess)
     // Possible forms are T[xxx], xxx[T], or xxx[keyof T], where T is a type variable
-    export interface IndexedAccessType extends TypeVariable {
+    export interface IndexedAccessType extends InstantiableType {
         objectType: Type;
         indexType: Type;
         constraint?: Type;
     }
 
     // keyof T types (TypeFlags.Index)
-    export interface IndexType extends Type {
-        type: TypeVariable | UnionOrIntersectionType;
+    export interface IndexType extends InstantiableType {
+        type: InstantiableType | UnionOrIntersectionType;
+    }
+
+    // T extends U ? X : Y (TypeFlags.Conditional)
+    export interface ConditionalType extends InstantiableType {
+        checkType: Type;
+        extendsType: Type;
+        trueType: Type;
+        falseType: Type;
+        /* @internal */
+        target?: ConditionalType;
+        /* @internal */
+        mapper?: TypeMapper;
+    }
+
+    // Type parameter substitution (TypeFlags.Substitution)
+    // Substitution types are created for type parameter references that occur in the true branch
+    // of a conditional type. For example, in 'T extends string ? Foo<T> : Bar<T>', the reference to
+    // T in Foo<T> is resolved as a substitution type that substitutes 'string & T' for T. Thus, if

```

Comment:
One of the problems that we discussed a few weeks back was that for something like

```ts
type ElementType<T> = T extends any[] ? T[number]: never
```

`T` would be viewed as `T & any[]` in the true branch type, which would result in something like `any & T[number]` which really just becomes `any`. If I understand correctly, I think key thing to inform other contributors about what you've done here is that substitution types are really just used for extra validation, and are ignored when a type is requested from `getConditionalType`.

I don't think that's obvious from just looking at this (you have to jump to `eraseSubstitutionType`), so I would try to mention this explicitly. Maybe something like

```ts
// Substitution types are created for type parameter references that occur in the true branch
// of a conditional type, but are *only* used for validation against type parameters and type operators.
```

etc.